[SofaKernel] Check if Quaternion has norm 1

olivier-goury opened the pull request:
Hello,
For the case of Rigid Mechanical Object (thus with position marked by the combination of a Vec3D and a quaternion for rotations), this PR adds a check if the quaternion is of unit 1 and normalizes it, if it is not. 
This PR does not handle the case when the user changes the value of the quaternion in the GUI.

Cheerio






______________________________________________________
<!--- Please leave this at the end of your message -->
This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

olivier-goury committed:
For RIgid Mechanical Object, add check if quaternion does have a unit norm and normalise it if not.

guparan commented on SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl:
```diff
@@ -1276,6 +1276,39 @@ void MechanicalObject<DataTypes>::reinit()
     if (translation.getValue()[0]!=0.0 || translation.getValue()[1]!=0.0 || translation.getValue()[2]!=0.0)
         this->applyTranslation( translation.getValue()[0],translation.getValue()[1],translation.getValue()[2]);
+
+    if (x.getValue()[0].size() == 7)   // The object is a Rigid3D.
```
Better way to check if Rigid3D ?

damienmarchal replied:
To answer your question, as this is templated code the right approach is probably to rely on code specialization instead of type branching. 

Eg:
```cpp
template<A>
MechanicalObject<A>::reinit()
{
     commonInitAllDataType();
     specificInitDependingOnData();
}

template<class A>
MechanicalObject<A>::commonInitAllDataType() { ....blahblha... }

template<class A>
MechanicalObject<A>::specificInitDependingOnData() {} /// Nothing by default
  
template<> /// Specialization on rigid2
MechanicalObject<Rigid3D>::specificInitDependingOnData() 
{
      ///put here the code related to quaternion normalization
} 
```

guparan commented on SofaKernel/modules/SofaBaseMechanics/MechanicalObject.inl:
```diff
@@ -1276,6 +1276,39 @@ void MechanicalObject<DataTypes>::reinit()
+        double epsilon = 1.0e-10;
+        for (size_t i = 0; i<x.getValue().size();i++)
+        {
+            double x3 = (double)x.getValue()[i][3];
```
Use `Quater::normalize` :+1:

damienmarchal commented:
Hello, 
so after a quick look at the class the fix to do is not in MechanicalObject but probably in Quater.inl, when the quaternion is converted from string to quaternion:
```cpp
 /// write to an output stream
    inline friend std::ostream& operator << ( std::ostream& out, const Quater& v )
    {
        out<<v._q[0]<<" "<<v._q[1]<<" "<<v._q[2]<<" "<<v._q[3];
        return out;
    }
    /// read from an input stream
    inline friend std::istream& operator >> ( std::istream& in, Quater& v )
    {
        in>>v._q[0]>>v._q[1]>>v._q[2]>>v._q[3];
        return in;
    }
```
This is sloppy code that should be made much more robust regarding the input.

Instead of reading directly the value, they should be buffered in double, then passed to a function like
this->setFromRawValue(); 
The checking code and normalization should be in this function.

Doing that in Quater.inl has the advantage that all quaternion read from string will always be checked and valided. 

The drawback (but this is a general drawback in how we handle error with message instead of using exception is how do we report error messages so that user can fix their scene.)

olivier-goury committed:
Update Quaternion default Normalisation. Does not work when user changes value in the GUI.

olivier-goury commented:
I want to make this PR less ambitious by solving the case only when the user loads a scene with a wrong quaternion. 
This does not solve the problem when the user modifies the quaternion wrongly in the GUI.
I let cleverer people fix this issue in the best way...

damienmarchal commented:
Quick summary:
Problem 1: 
Quaternion (when unnormalized) can be used to encode more than juste R3 orientation so we shouldn't force the normalization in the Quaternion operator>>().
We should have normalization detection and renormalization done only on object that we know for sure are R3 orientation. So to me the proper fix should be done in the operator>> of Rigid3Type.

Problem 2:
About reporting warning to the users. This cannot be done in the Quater::normalize function. Because normalization is a perfectly valid operation to apply on an un-normalized quaternion and thus there is no meaning to report a warning. Actually it is like printing  a warning when we are doing exactly what the function is supposed to do. 

Now, the problem goes far beyond this PR as reporting error/exception is a complex issue in Sofa as we rely on 'message'  instead of exception and C-style error handling vs exception vs printing (as in sofa) is an as old problem as c++ and in our code base all the approaches have been used.

olivier-goury committed:
Update so that quaternion normalization is performed within the RigidType class, i.e a quarternion has to be normalised when it is used in a Rigid3d object.

damienmarchal commented:
Read... so merge
